lib/ecto/association.ex

```diff
@@ -272,8 +272,8 @@ defmodule Ecto.Association do
       ** (ArgumentError) association :comments_v1 queryable must be a schema, a {source, schema}, or a query. got: "wrong"
   """
   def related_from_query(atom, _name) when is_atom(atom), do: atom
-  def related_from_query({source, schema}, _name) when is_binary(source) and is_atom(schema), do: schema
-  def related_from_query(%Ecto.Query{from: {source, schema}} = query, name) when is_binary(source) and is_atom(schema) do
+  def related_from_query(%FromExpr{source: source, schema: schema}, _name) when is_binary(source) and is_atom(schema), do: schema
+  def related_from_query(%Ecto.Query{from: %FromExpr{source: source, schema: schema}} = query, name) when is_binary(source) and is_atom(schema) do
```

Comment:
Regarding error message at https://github.com/elixir-ecto/ecto/pull/2497/files#diff-f547e032c73d51297b726338f16756c5R284 - I have left it unchanged b/c I assume it's a hint for the end user who will input and see external representation of `from`, which still is `{source, schema}`.